src/qibosoq/programs/base.py

```diff
@@ -272,11 +272,28 @@ def collect_shots(self) -> Tuple[list, list]:
 
         unique_adcs, adc_count = np.unique(adcs, return_counts=True)
 
+        number_of_adcs = len(unique_adcs)
+        len_buf = len(self.di_buf[0])
+
         for idx, adc_ch in enumerate(unique_adcs):
-            count = adc_count[idx]
-            shape = (count, self.expts, self.reps) if self.expts else (count, self.reps)
-            i_val = self.di_buf[idx].reshape(shape) / lengths[idx]
-            q_val = self.dq_buf[idx].reshape(shape) / lengths[idx]
+            count = int(adc_count[idx])
+            try:
+                shape = (
+                    count,
+                    int(np.prod(self.sweep_axes)),
+                    self.reps,
+                )  # type: Union[Tuple[int, int], Tuple[int, int, int]]
+            except AttributeError:
+                shape = (count, self.reps)
+
+            i_val = (
+                self.di_buf[idx][: len_buf // number_of_adcs].reshape(shape)
+                / lengths[idx]
+            )
+            q_val = (
+                self.dq_buf[idx][: len_buf // number_of_adcs].reshape(shape)
+                / lengths[idx]
+            )
```

Comment:
I know you're delivering `tot_i` and `tot_q`, but it would be nicer to stack them in a single array (with one more dimension, of length 2), such that you won't have to repeat the same operations.

Even if you wanted to keep returning the same values, you could also stack just internally in this function, at least for the time being (it would be a nice compromise for this PR, in such a way that the diff remains limited).

Reply:
For now, I am satisfied of what we have now. I can open a PR to fix all these shape-related problems for results.
I am using different arrays also since I am using di_buf and dq_buf. However I know that there is also a dd_buf that could be helpful, but I have to look into that more

Reply:
Actually, having a look from a broader perspective, you could just stack `self.di_buf` and `self.dq_buf` even outside the loop, that could be kept in NumPy.

The only weird thing is that the shape depends on `idx`...

Is it really possible that you have different ranks? Or are they going to be different just in `counts`?

Otherwise, you could do your processing (divide by `lengths`, cut by `len_buf // number_of_adcs]`) and just reshape at the end.

Reply:
As a quick reply (I could refine the code):
```python
    stacked = np.stack((self.di_buf, self.dq_buf), axis=0)[:, :, :len_acq] / lengths[idx]
    shapes = []
    for idx, adc_ch in enumerate(unique_adcs):
        count = int(adc_count[idx])
        try:
            shapes.append((count, int(np.prod(self.sweep_axes)), self.reps))
        except AttributeError:
            shapes.append((count, self.reps))

    # something like stacked.reshape(shapes)
    return tot_i, tot_q
```

But if you have to really `.reshape()` with different `count`s, the `.reshape()` could be kept in the loop.

Reply:
Sorry, also lengths[idx] could change so it should be kept inside the loop.
Honestly I am not seeing much more clarity, I think that could came if we use the dd_buf, but otherwise it seems to become just more convoluted without much advantage

Reply:
Whoops, I wrote it too fast.

However, it is fine for `lengths[idx]` to depend on `idx`, since it is just an array.
You can both do:
```py
x = np.ones((20, 10)) * 10
y = np.ones((20, 10)) * np.arange(10)
```
NumPy operations are applied element-wise (here you should adequately compensate for the higher rank with `np.newaxis`, but that's it).

Arrays operations are more expressive, and thus more declarative: you are closer to describe the result you want, rather than the algorithm to obtain.
When applied in Python, they are also faster (just because of Python's slow-down), though I hope here is not a problem.

Reply:
However, what is clearer is definitely subjective. I'm just telling you my experience, and motivating it.
If eventually you like more the current one because you believe there is no better way, that's just the end of the story :)

Reply:
This would be my proposal:
```py
    tot = []
    stacked = np.stack((self.di_buf, self.dq_buf)) / lengths[:, np.newaxis]
    for idx, count in enumerate(adc_count.astype(int)):
        try:
            shape = (count, int(np.prod(self.sweep_axes)), self.reps) 
        except AttributeError:
            shape = (count, self.reps)

		tot.append(stacked[:, idx].reshape(shape))

    return list(zip(*tot))
```

Such that the `for` loop is only restricted to the only thing that you couldn't vectorize (i.e. the reshape, because of the different shapes).

Reply:
I tried to implement it, but I cannot make mypy happy about the return type (that it sees as tuple[list[any], ...], while ti should be tuple[list, list]). So I added a mypy exception.
I am using lists to send them through the socket (if I remember correctly I cannot do otherwise).

I am still not fully convinced, but I'm trusting you :-)

Reply:
I tested the output, but in pseudoinput.

If you could provide real input, I would check whether the output is the same before and after.